File: scripts/write_answer.py
#!/usr/bin/env python3
"""
write_answer.py

PURPOSE: Helper script for agents to write answers to the requirements file

This script simplifies the process of updating the answers JSON file during
requirements gathering. It loads existing answers, adds/updates one answer,
and writes back to the file.

USAGE:
  python write_answer.py <answers-file> <question-id> <answer>

EXAMPLES:
  # Text answer
  python write_answer.py requirements.json systems "Salesforce to Slack"

  # Choice answer
  python write_answer.py requirements.json trigger_type webhook

  # Multi-choice answer (JSON array)
  python write_answer.py requirements.json error_handling '["Automatic retry", "Log errors"]'

EXIT CODES:
  0 - Success
  1 - Error (missing arguments, invalid JSON, etc.)
"""

import json
import sys


def main():
    if len(sys.argv) < 4:
        print("❌ Usage: python write_answer.py <answers-file> <question-id> <answer>")
        return 1

    answers_file = sys.argv[1]
    question_id = sys.argv[2]
    answer_raw = sys.argv[3]

    # Try to parse answer as JSON (for arrays/objects), fall back to string
    try:
        answer = json.loads(answer_raw)
    except json.JSONDecodeError:
        answer = answer_raw

    # Load existing answers
    try:
        with open(answers_file, "r") as f:
            answers = json.load(f)
    except FileNotFoundError:
        answers = {}
    except json.JSONDecodeError as e:
        print(f"❌ Invalid JSON in {answers_file}: {e}", file=sys.stderr)
        return 1

    # Add/update answer
    answers[question_id] = answer

    # Validation warning for connection-type questions
    connection_questions = [
        "source_connection_type",
        "destination_connection_type",
    ]

    if question_id in connection_questions:
        if isinstance(answer, dict):
            if "inputs" not in answer or not answer.get("inputs"):
                print("", file=sys.stderr)
                print(
                    "⚠️  WARNING: Connection answer is missing 'inputs' array!",
                    file=sys.stderr,
                )
                print(
                    "   This will cause 'No credentials needed' error later.",
                    file=sys.stderr,
                )
                print("", file=sys.stderr)
                print(
                    "   Expected: Full object from choice 'value' field with:",
                    file=sys.stderr,
                )
                print("   - key", file=sys.stderr)
                print("   - label", file=sys.stderr)
                print("   - auth_type", file=sys.stderr)
                print("   - required_inputs (array)", file=sys.stderr)
                print("   - inputs (array) ← CRITICAL FOR CREDENTIALS", file=sys.stderr)
                print("", file=sys.stderr)
        else:
            print("", file=sys.stderr)
            print(
                "⚠️  WARNING: Connection answer should be an object, not string!",
                file=sys.stderr,
            )
            print(
                "   Use the full 'value' object from the choice, not just label.",
                file=sys.stderr,
            )
            print("", file=sys.stderr)

    # Write back
    try:
        with open(answers_file, "w") as f:
            json.dump(answers, f, indent=2)
        print(f"✅ Answer written to {answers_file}")
        print(f"   {question_id} = {answer}")
        return 0
    except Exception as e:
        print(f"❌ Failed to write file: {e}", file=sys.stderr)
        return 1


if __name__ == "__main__":
    sys.exit(main())

File: scripts/write-answer.ts
#!/usr/bin/env npx tsx
/**
 * write-answer.ts
 *
 * Helper script for agents to write answers to the requirements file.
 *
 * USAGE:
 *   npx tsx write-answer.ts <answers-file> <question-id> <answer>
 *   npx tsx write-answer.ts <answers-file> --flow <flow-id> <question-id> <answer>
 *   echo '<answer>' | npx tsx write-answer.ts <answers-file> [--flow <flow-id>] <question-id>
 *
 * When --flow is provided, the answer is written under answers.flows[flowId].
 * When omitted, the answer is written at the root level (backward compatible).
 * When the answer argument is omitted, the script reads from stdin.
 *
 * EXIT CODES:
 *   0 - Success
 *   1 - Error
 */

import { readFileSync, writeFileSync } from "node:fs";

function main(): number {
  const args = process.argv.slice(2);

  if (args.length < 2) {
    console.log(
      "Usage: npx tsx write-answer.ts <answers-file> [--flow <flow-id>] <question-id> <answer>\n" +
      "       echo '<answer>' | npx tsx write-answer.ts <answers-file> [--flow <flow-id>] <question-id>"
    );
    return 1;
  }

  const answersFile = args[0];

  // Parse --flow flag
  let flowId: string | null = null;
  let restArgs: string[];

  if (args[1] === "--flow") {
    if (args.length < 4) {
      console.error("--flow requires a flow ID and question ID");
      return 1;
    }
    flowId = args[2];
    restArgs = args.slice(3);
  } else {
    restArgs = args.slice(1);
  }

  const questionId = restArgs[0];
  let answerRaw: string | undefined = restArgs[1];

  if (answerRaw === "--json" && restArgs.length > 2) {
    answerRaw = restArgs[2];
  }

  // If no answer argument, read from stdin
  if (answerRaw === undefined) {
    try {
      answerRaw = readFileSync(0, "utf-8").trim();
    } catch {
      console.error("No answer argument provided and failed to read from stdin");
      return 1;
    }
    if (!answerRaw) {
      console.error("No answer argument provided and stdin is empty");
      return 1;
    }
  }

  // Try to parse answer as JSON (for arrays/objects), fall back to string
  let answer: unknown;
  try {
    answer = JSON.parse(answerRaw);
  } catch {
    answer = answerRaw;
  }

  // Load existing answers
  let answers: Record<string, unknown>;
  try {
    const content = readFileSync(answersFile, "utf-8");
    answers = JSON.parse(content);
  } catch (e: unknown) {
    if ((e as NodeJS.ErrnoException).code === "ENOENT") {
      answers = {};
    } else {
      console.error(`Invalid JSON in ${answersFile}: ${e}`);
      return 1;
    }
  }

  // Determine target: root or flows[flowId]
  let target: Record<string, unknown>;
  if (flowId) {
    if (!answers.flows || typeof answers.flows !== "object") {
      answers.flows = {};
    }
    const flows = answers.flows as Record<string, Record<string, unknown>>;
    if (!flows[flowId] || typeof flows[flowId] !== "object") {
      flows[flowId] = {};
    }
    target = flows[flowId];
  } else {
    target = answers;
  }

  // Add/update answer
  target[questionId] = answer;

  // Validation warning for connection-type questions
  const connectionQuestions = [
    "source_connection_type",
    "destination_connection_type",
  ];

  if (connectionQuestions.includes(questionId)) {
    if (typeof answer === "object" && answer !== null) {
      const obj = answer as Record<string, unknown>;
      if (!obj.inputs || (Array.isArray(obj.inputs) && obj.inputs.length === 0)) {
        console.error("");
        console.error("WARNING: Connection answer is missing 'inputs' array!");
        console.error("   This will cause 'No credentials needed' error later.");
        console.error("");
        console.error("   Expected: Full object from choice 'value' field with:");
        console.error("   - key");
        console.error("   - label");
        console.error("   - auth_type");
        console.error("   - required_inputs (array)");
        console.error("   - inputs (array) <- CRITICAL FOR CREDENTIALS");
        console.error("");
      }
    } else {
      console.error("");
      console.error("WARNING: Connection answer should be an object, not string!");
      console.error("   Use the full 'value' object from the choice, not just label.");
      console.error("");
    }
  }

  // Write back
  try {
    writeFileSync(answersFile, JSON.stringify(answers, null, 2));
    const prefix = flowId ? `[flow: ${flowId}] ` : "";
    console.log(`${prefix}Answer written to ${answersFile}`);
    console.log(`   ${questionId} = ${typeof answer === "string" ? answer : JSON.stringify(answer)}`);
    return 0;
  } catch (e) {
    console.error(`Failed to write file: ${e}`);
    return 1;
  }
}

process.exit(main());
